src/sinks/util/encoding/adapter.rs

```diff
@@ -0,0 +1,335 @@
+#![deny(missing_docs)]
+
+use super::{EncodingConfiguration, TimestampFormat};
+use crate::{
+    codecs::encoding::{Framer, FramingConfig, Serializer, SerializerConfig},
+    event::{Event, PathComponent},
+};
+use core::fmt::Debug;
+use serde::{Deserialize, Serialize};
+use std::marker::PhantomData;
+
+/// Trait used to migrate from a sink-specific `Codec` enum to the new
+/// `Box<dyn FramingConfig>`/`Box<dyn SerializerConfig>` encoding configuration.
+pub trait EncodingConfigMigrator {
+    /// The sink-specific encoding type to be migrated.
+    type Codec;
+
+    /// Returns the framing/serializer configuration that is functionally equivalent to the given
+    /// legacy codec.
+    fn migrate(codec: &Self::Codec) -> (Option<Box<dyn FramingConfig>>, Box<dyn SerializerConfig>);
+}
+
+/// This adapter serves to migrate sinks from the old sink-specific `EncodingConfig<T>` to the new
+/// `Box<dyn FramingConfig>`/`Box<dyn SerializerConfig>` encoding configuration - while keeping
+/// backwards-compatibility.
+#[derive(Debug, Clone, Serialize, Deserialize)]
+#[serde(untagged)]
+pub enum EncodingConfigAdapter<
+    LegacyEncodingConfig: EncodingConfiguration + Debug + Clone + 'static,
+    Migrator: EncodingConfigMigrator<Codec = <LegacyEncodingConfig as EncodingConfiguration>::Codec>
+        + Debug
+        + Clone,
+> {
+    /// The encoding configuration.
+    Encoding(EncodingConfig),
+    /// The legacy sink-specific encoding configuration.
+    LegacyEncodingConfig(LegacyEncodingConfigWrapper<LegacyEncodingConfig, Migrator>),
+}
+
+impl<
+        LegacyEncodingConfig: EncodingConfiguration + Debug + Clone + 'static,
+        Migrator: EncodingConfigMigrator<Codec = <LegacyEncodingConfig as EncodingConfiguration>::Codec>
+            + Debug
+            + Clone,
+    > EncodingConfigAdapter<LegacyEncodingConfig, Migrator>
```

Comment:
These would probably be easier to read if you moved the trait bounds to a `where` instead of being inline.

Reply:
Addressed in 9f48092762fe600c09be618a0c4175c06492b636.